JDK-8261079: Fix support for @hidden in classes and interfaces

hns opened the pull request:
This change improves support for the `@hidden` tag to suppress documentation for specific elements, especially in the context of classes, interfaces and inheritance/method overriding.

The important changes are in `Utils` and in `VisibleMemberTable`. (There is also a number of non-related small code cleanup changes, sorry about that, I know it makes review a bit harder but I couldn't resist.)

In `Utils` the most important change are:

 - Consider types as "undocumented enclosures" if they are marked with a `@hidden` tag
 - Check for `@hidden` tags even in non-included elements as they may be included via undocumented enclosures
 - To counter the above change, only run doclint on elements that are either included or contained in an included type, as we don't want to report warnings or errors for non-included code. 
 
In `VisibleMemberTable` there is a subtle change to not consider an overriding method as a "simple override" if the overridden method is hidden or otherwise invisible but in turn is a non-simple override of a method further up the inheritance chain. This resulted in methods which should have been documented as declared locally to be documented as declared in supertypes. I also did a bit of renaming in `VisibleMemberTable` to make the purpose of things a bit clearer.

Other than that, most of the changes consist of added calls to `utils.hasHiddenTag(Element)`, usually with the purpose of not generating links to things that are not there.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261079](https://bugs.openjdk.java.net/browse/JDK-8261079): Fix support for @hidden in classes and interfaces


### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to 66e9b54a0f60e85fbb18d9a316742e3f2640ebdb


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2406/head:pull/2406`
`$ git checkout pull/2406`

hns commented:
Since I changed the code for handling method overrides in `VisibleMemberTable` I compared the generated documentation for current JDK 17 with and without this patch. There were no changes in the generated documentation.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2788,6 +2793,19 @@ public void checkJavaScriptInOption(String name, String value) {
+        if (isOverviewElement(e) || isIncluded(e)) {
+            return true;
+        }
+        // Run doclint on non-incuded members of included type elements.
```
typo `incuded`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2723,7 +2726,9 @@ public DocCommentTree getDocCommentTree0(Element element) {
                 }
                 // run doclint even if docCommentTree is null, to trigger checks for missing comments
-                configuration.runDocLint(path);
+                if (shouldRunDocLint(element)) {
+                    configuration.runDocLint(path);
+                }
```
I'm disappointed there is so much disruption to the code because of the enhanced treatment of `@hidden`.Maybe it's necessary, but I was hoping/expecting that we could add a separate code path for the specific purpose of seeing if a non-selected(?) class has a `@hidden` tag. 

For example, leave all the code for block tags and doc comment tree alone unchanged, and then provide `isHidden(Element)`  by just using `getDocCommentInfo`, which gives cached access to the `DocCommentTree` then scan the block tags for a `HIDDEN` tree.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -258,7 +258,7 @@ public ExecutableElement getSimplyOverriddenMethod(ExecutableElement e) {
         OverridingMethodInfo found = overriddenMethodTable.get(e);
         if (found != null && found.simpleOverride) {
-            return found.overrider;
+            return found.overridden;
```
changing from `overrider` to `overridden` seems like a pretty big change, and not just a terminology cleanup.

Is this related to the change you described in the introduction?

jonathan-gibbons reviewed (changes requested):
Two primary areas:

1. I think the changes to Utils to determine if a tag is hidden can be improved/simplified
2. I'm trying to understand the changes to `VisibleMemberTable`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -1553,15 +1555,16 @@ public String propertyName(ExecutableElement e) {
      * @return true if it exists, false otherwise
      */
     public boolean hasHiddenTag(Element e) {
-        // prevent needless tests on elements which are not included
-        if (!isIncluded(e)) {
+        // prevent needless tests on elements which are neither included nor selected.
+        // Non-included members may still be visible via "transclusion" from undocumented enclusure
+        if (!isIncluded(e) && !configuration.docEnv.isSelected(e)) {
```
This comment applies to the whole of this method body.

I think the `@hidden` (or @treatAsprivate`) tag is sufficiently special that it should be handled separately, especially in the cases where the containing comment would not otherwise be read, because the element is not included/selected. See other/earlier comments about avoiding `getDocComment` and going to the underlying `getDocCommentInfo`, which will (should?) completely bypass `getDocComment` and (importantly) the use of doclint.

hns commented:
Thanks for the review, Jon! 

I pushed a new commit that should address your concerns. For the lookup of `@hidden` tags in non-included elements I added a new `hasBlockTagUnchecked` method that avoids caching and doclint checking by using  `getDocCommentInfo` as you suggested. Note that this isn't used for the javafx "treatAsPrivate" tag as changing the semantics of an undocumented/internal tag didn't seem like a good idea on second thought.

Regarding the renaming of `overrider` to `overridden`: the old name was a misnomer, the new name actually reflects the meaning of the field as it contains the overridden method, not the overriding one. I actually went the whole way and renamed the nested class as well, from `OverridingMethodInfo` to `OverriddenMethodInfo`.

hns commented:
/integrated

hns commented:
/integrate
